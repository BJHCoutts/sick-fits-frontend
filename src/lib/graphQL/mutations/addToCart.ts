import { gql } from "@apollo/client"

export const ADD_TO_CART_MUTATION = gql`
	mutation ADD_TO_CART($productId: ID!) {
		addToCart(productId: $productId) {
			id
		# quantity
		# product {
		# 	id
		# 	name
		# }
		# user {
		# 	id
		# 	name
		# 	email
			# }
		}
	}
`