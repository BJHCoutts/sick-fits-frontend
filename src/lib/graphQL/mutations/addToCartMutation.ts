import { gql } from "@apollo/client"

export const ADD_TO_CART_MUTATION = gql`
	mutation ADD_TO_CART_MUTATION($productId: ID!) {
		addToCart(productId: $productId) {
			id
		}
	}
`