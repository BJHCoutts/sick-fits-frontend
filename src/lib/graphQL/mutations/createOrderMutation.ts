import { gql } from "@apollo/client"

export const CREATE_ORDER_MUTATION = gql`
	mutation CREATE_ORDER_MUTATION($token:String!) {
  checkout (token:$token) {
    id
    charge
    total
    items{
      id
      name
    }
  }
}
`