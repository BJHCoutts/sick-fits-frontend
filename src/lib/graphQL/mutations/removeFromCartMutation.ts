import { gql } from "@apollo/client"

export const REMOVE_FROM_CART_MUTATION = gql`
	mutation REMOVE_FROM_CART_MUTATION ($id:ID!){
		deleteCartItem (id:$id) {
			id
			quantity
			product {
				name
			}
		}
	}
`