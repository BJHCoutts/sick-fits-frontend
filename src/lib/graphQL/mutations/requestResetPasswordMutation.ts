import { gql } from "@apollo/client"

export const REQUEST_RESET_PASSWORD_MUTATION = gql`
	mutation REQUEST_RESET_PASSWORD_MUTATION($email: String!) {
		sendUserPasswordResetLink (email:$email) {
      code
      message
    }
	}
`