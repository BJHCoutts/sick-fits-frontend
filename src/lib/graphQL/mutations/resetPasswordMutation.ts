import { gql } from "@apollo/client"

export const RESET_PASSWORD_MUTATION = gql`
	mutation RESET_PASSWORD_MUTATION(
		$email: String!, 
		$password: String!, 
		$token: String!
	) {
		redeemUserPasswordResetToken(
			email: $email
			password: $password
			token: $token
		) {
			code
			message
		}
	}
`