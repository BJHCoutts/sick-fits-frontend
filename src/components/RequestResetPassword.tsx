import { useMutation } from '@apollo/client'
import useForm from '../lib/functions/useForm'
import SForm from './styles/SForm'
import DisplayError from './ErrorMessage'
import { REQUEST_RESET_PASSWORD_MUTATION } from '../lib/graphQL/mutations/requestResetPasswordMutation'
import Link from 'next/link'

export default function RequestResetPassword() {
	const { inputs, handleChange, resetForm } = useForm({
		email: '',
	})

	const [resetPassword, { data, loading, error }] = useMutation(
		REQUEST_RESET_PASSWORD_MUTATION,
		{
			variables: inputs,
		}
	)

	async function handleSubmit(e: React.FormEvent) {
		e.preventDefault()
		const res = await resetPassword().catch(console.error)
		resetForm()
	}

	return (
		<>
			<SForm method="POST" onSubmit={handleSubmit}>
				<h2>Request Password Reset</h2>
				<DisplayError error={error} />
				<fieldset disabled={loading} aria-busy={loading}>
					{data?.sendUserPasswordResetLink === null && (
						<p>
							Success! Password has been changed. Enter the ?token=token query
							string into the URL <Link href={`/reset`}>here</Link>
						</p>
					)}
					<label htmlFor="email">
						Email
						<input
							type="email"
							name="email"
							placeholder="Your Email Address"
							autoComplete="email"
							value={inputs.email}
							onChange={handleChange}
						/>
					</label>
					<button type="submit">Request Password Reset</button>
				</fieldset>
			</SForm>
		</>
	)
}
