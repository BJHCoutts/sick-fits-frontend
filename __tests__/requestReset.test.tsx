import { render, screen } from '@testing-library/react'
import { REQUEST_RESET_PASSWORD_MUTATION } from '../src/lib/graphQL/mutations/requestResetPasswordMutation'
import { MockedProvider } from '@apollo/client/testing'
import RequestResetPassword from '../src/components/RequestResetPassword'
import userEvent from '@testing-library/user-event'

const email = 'test@example.com'
const mocks = [
	{
		request: {
			query: REQUEST_RESET_PASSWORD_MUTATION,
			variables: { email },
		},
		result: {
			data: { sendUserPasswordResetLink: null },
		},
	},
]

describe('<RequestResetPassword />', () => {
	it('renders and matches snapshot', () => {
		const { container } = render(
			<MockedProvider>
				<RequestResetPassword />
			</MockedProvider>
		)
		expect(container).toMatchSnapshot()
	})

	it(' calls the mutation when submitted', async () => {
		const { container, debug } = render(
			<MockedProvider mocks={mocks}>
				<RequestResetPassword />
			</MockedProvider>
		)

		// debug(screen.getByLabelText('Email'))
		const emailInput = screen.getByLabelText('Email')
		userEvent.type(emailInput, email)

		debug(emailInput)
		// expect(screen.getByDisplayValue(email)).toBeInTheDocument()

		// userEvent.click(screen.getByTestId('request-reset-password-button'))

		// const success = await screen.findByText(/success/i)
		// debug()
		// TODO Success message not being triggered in test, investigate why
		// await waitFor(() => {
		// 	expect(success).toBeInTheDocument()
		// })
		// expect(success).toBeInTheDocument()
	})
})
