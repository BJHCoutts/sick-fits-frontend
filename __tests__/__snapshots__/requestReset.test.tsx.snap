// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestResetPassword /> renders and matches snapshot 1`] = `
<div>
  <form
    class="sc-bdfCDU cdIIGl"
    method="POST"
  >
    <h2>
      Request Password Reset
    </h2>
    <fieldset
      aria-busy="false"
    >
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          id="email"
          name="email"
          placeholder="Your Email Address"
          type="email"
          value=""
        />
      </label>
      <button
        data-testid="request-reset-password-button"
        type="submit"
      >
        Request Password Reset
      </button>
    </fieldset>
  </form>
</div>
`;
